server/datastore/mysql/microsoft_mdm.go

```diff
@@ -135,210 +128,214 @@ func (ds *Datastore) MDMWindowsDeleteEnrolledDeviceWithDeviceID(ctx context.Cont
+			case fleet.CmdResults:
+				uuidsToResults[*cmdRef] = protoOp.Cmd
+				cmdUUIDs = append(cmdUUIDs, *cmdRef)
+			}
+		}
 
-	query := `
-        SELECT
-            command_uuid,
-            device_id,
-            session_id,
-            message_id,
-            command_id,
-            cmd_verb,
-            setting_uri,
-            setting_value,
-            system_origin,
-            rx_error_code,
-            rx_cmd_result,
-            created_at,
-            updated_at
-        FROM
-            old_windows_mdm_commands
-        WHERE
-            device_id = ?
-    `
+		// no relevant commands to tracks is a noop
+		if len(cmdUUIDs) == 0 {
+			return nil
+		}
 
-	if err := sqlx.SelectContext(ctx, ds.reader(ctx), &commands, query, deviceID); err != nil {
-		return nil, ctxerr.Wrap(ctx, err, "get Windows MDM commands by device id")
-	}
+		// store the full response
+		sqlResult, err := tx.ExecContext(ctx, saveFullRespStmt, enrollment.ID, fullResponse.Raw)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "saving full response")
+		}
+		responseID, _ := sqlResult.LastInsertId()
 
-	return commands, nil
+		// find commands we sent that match the UUID responses we've got
+		stmt, params, err := sqlx.In(findCommandsStmt, cmdUUIDs)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "building IN to search matching commands")
+		}
+		var matchingUUIDs []string
+		err = sqlx.SelectContext(ctx, tx, &matchingUUIDs, stmt, params...)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "selecting matching commands")
+		}
+
+		// TODO(roberto): no matching commands is a noop? should we
+		// store the results anyway? The full response is already
+		// stored at this point if we need to retrieve those at some point.
+		if len(matchingUUIDs) == 0 {
+			ds.logger.Log("warn", "unmatched commands", "uuids", cmdUUIDs)
+			return nil
+		}
+
+		// for all the matching UUIDs, try to find any <Status> or
+		// <Result> entries to track them as responses.
+		var args []any
+		var sb strings.Builder
+		for _, uuid := range matchingUUIDs {
+			statusCode := ""
+			if status, ok := uuidsToStatus[uuid]; ok && status.Data != nil {
+				statusCode = *status.Data
+			}
+
+			rawResult := []byte{}
+			if result, ok := uuidsToResults[uuid]; ok && result.Data != nil {
+				rawResult = []byte(*result.Data)
+			}
+			args = append(args, enrollment.ID, uuid, rawResult, responseID, statusCode)
+			sb.WriteString("(?, ?, ?, ?, ?),")
+		}
+
+		// store the command results
+		stmt = fmt.Sprintf(insertResultsStmt, strings.TrimSuffix(sb.String(), ","))
+		if _, err = tx.ExecContext(ctx, stmt, args...); err != nil {
+			return ctxerr.Wrap(ctx, err, "inserting command results")
+		}
+
+		// dequeue the commands
+		// TODO(roberto): is this  the right place to dequeue the commands?
+		//
+		// If we didn't immediately get a response for a command we
+		// already sent, and we don't dequeue it, it's going to be sent
+		// again.  Is this what we want?
```

Comment:
Makes sense to me to stop sending a command from the moment we get a response. Is that roughly how the Apple nano mdm package works?

Reply:
that's correct! I think it's assuming you'll get a response from the device before you have the chance to queue more commands, I wonder if we can assume the same for Windows @marcosd4h ?

Reply:
On Windows, you will get a response for a command sent to a device on the following SyncML message sent from the device to the server. It should be safe to stop sending a command when server gets the response.

Reply:
thanks! I will remove this TODO